From 0d7c72f446903c64924ebdc62834cab07dcada9a Mon Sep 17 00:00:00 2001
From: Alexandros Frantzis <alexandros.frantzis@collabora.com>
Date: Wed, 27 Oct 2021 17:41:37 +0300
Subject: [PATCH 058/147] winewayland.drv: Update children surfaces when parent
 is updated

When a parent Wayland surface is recreated, mark the children surfaces
as requiring an update, so that the get eventually recreated using the
new parent.

Signed-off-by: Alexandros Frantzis <alexandros.frantzis@collabora.com>
---
 dlls/winewayland.drv/window.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/dlls/winewayland.drv/window.c b/dlls/winewayland.drv/window.c
index cf66b86151d..0bd27f82b0c 100644
--- a/dlls/winewayland.drv/window.c
+++ b/dlls/winewayland.drv/window.c
@@ -514,7 +514,26 @@ static void wayland_win_data_update_wayland_surface(struct wayland_win_data *dat
     if (data->wayland_surface != surface)
     {
         if (data->wayland_surface)
+        {
+            struct wayland_surface *child;
+
+            /* Dependent Wayland surfaces require an update, so that they point
+             * to the updated surface. */
+            EnterCriticalSection(&data->wayland_surface->crit);
+            wl_list_for_each(child, &data->wayland_surface->child_list, link)
+            {
+                struct wayland_win_data *child_data;
+                if ((child_data = wayland_win_data_get(child->hwnd)))
+                {
+                    child_data->wayland_surface_needs_update = TRUE;
+                    wayland_win_data_release(child_data);
+                }
+            }
+            LeaveCriticalSection(&data->wayland_surface->crit);
+
             wayland_surface_unref(data->wayland_surface);
+        }
+
         data->wayland_surface = surface;
     }
 }
-- 
2.34.1

